pkg/apis/wsecurity/v1/count.go

```diff
@@ -0,0 +1,146 @@
+package v1
+
+import (
+	"bytes"
+	"fmt"
+)
+
+//////////////////// CountProfile ////////////////
+
+// Exposes ValueProfile interface
+type CountProfile uint8
+
+func (profile *CountProfile) Profile(args ...interface{}) {
+	*profile = CountProfile(args[0].(uint8))
+}
+
+func (profile *CountProfile) String(depth int) string {
+	return fmt.Sprintf("%d", uint8(*profile))
+}
+
+//////////////////// CountPile ////////////////
+
+// Exposes ValuePile interface
+type CountPile []uint8
+
+func (pile *CountPile) Add(valProfile ValueProfile) {
+	profile := *valProfile.(*CountProfile)
+	*pile = append(*pile, uint8(profile))
+}
+
+func (pile *CountPile) Clear() {
+	*pile = nil
+}
+
+func (pile *CountPile) Merge(otherValPile ValuePile) {
+	otherPile := otherValPile.(*CountPile)
+	*pile = append(*pile, *otherPile...)
```

Comment:
The documentation claims that `*otherPile` will be unusable after this operation, but it looks like it will be usable but _may alias some values with `pile`_.

Why not call `otherPile.Clear()` at the end of this?

Reply:
`but it looks like it will be usable but may alias some values with pile` - correct, but we should no longer use pile without creating new values in pile  - we should not change one pile via changing the other.

Think of the dataTypes as offering many layers where each layer is implemented by either basic data types or other layers. It is therefore enough to call Clear() on the top layer, rather than calling it in every piece of data in each layer. 

Calling `otherPile.Clear()` at the end of any implementation of Merge() will result in many redundant calls for `Clear()`. It may be better to leave this to the user of this package - the user may call `Clear()` once on the top level dataType after calling the top level `Merge()` or decide not to call it if it will no longer be used.

Reply:
The cost of calling `Clear` appears to be a single pointer assignment. The cost of accidentally re-using a `Pile` is silent data corruption.

If we _are_ concerned about performance here, I can go over the design with an eye to performance, and suggest some benchmarks to write. There are a number of cases here where it seemed like we weren't hugely performance-sensitive (e.g. not sorting the ranges in CountPiles).

Reply:
Agreed.  

Rough estimate - we are going to have some 1000 counters per session (Req-Resp)
Each counter will typically have zero or one or two ranges - no more. 
Yes, specifically on these widely used Counters we need to be extra careful with lines of code executed.

I think we can either:
- decide to DeepCopy always and not affect other items
- decide to Clear other items always
- decide to Clear other items only if we consume them such that we never reach silent data corruption
- decide to make only some DataTypes exposed  (e.g. avoid exposing low-level data types and do Clear() only on exposed DataTypes
- mark the other data type as "consumed" (not sure if this will make more sense than simply clearing them) 
 
also - we need to add a list of "Things to revisit" into the repo such that we do not lose sight of those.

Reply:
Added to TODO list